tests/nlu/featurizers/test_lexical_syntactic_featurizer.py

```diff
@@ -195,3 +195,93 @@ def test_text_featurizer_using_pos_with_action_text(
+                ],
+                ["prefix2", "suffix1", "suffix3", "pos", "upper"],
+            ],
+        )
+    ],
+)
+def test_text_featurizer_using_feature_to_idx_dict(
+    sentence: Text, featurizer_config: List[List[Text]], spacy_nlp
+):
+    featurizer = LexicalSyntacticFeaturizer({"features": featurizer_config})
+
+    train_message = Message(data={TEXT: sentence, ACTION_TEXT: sentence})
+    test_message = Message(data={TEXT: sentence, ACTION_TEXT: sentence})
+
+    train_message.set(SPACY_DOCS[TEXT], spacy_nlp(sentence))
+    train_message.set(SPACY_DOCS[ACTION_TEXT], spacy_nlp(sentence))
+    test_message.set(SPACY_DOCS[TEXT], spacy_nlp(sentence))
+    test_message.set(SPACY_DOCS[ACTION_TEXT], spacy_nlp(sentence))
+
+    SpacyTokenizer().process(train_message)
+    SpacyTokenizer().process(test_message)
+
+    featurizer.train(TrainingData([train_message]))
+    featurizer.process(test_message)
+
+    seq_vec, sen_vec = test_message.get_sparse_features(TEXT, [])
+    if seq_vec:
+        seq_vec = seq_vec.features
+    if sen_vec:
+        sen_vec = sen_vec.features
+
+    # Check that all requested features were indexed
+    window_center = len(featurizer_config) // 2
+    for index in range(len(featurizer_config)):
+        for feature in featurizer_config[index]:
+            assert (
+                f"{index - window_center}:{feature}" in featurizer.feature_to_idx_dict
+            )
+
+    # Check that all indexed features were requested
+    for feature in featurizer.feature_to_idx_dict.keys():
+        i, feature_function = feature.split(":")
+        index = int(i)
+        assert index + window_center < len(featurizer_config)
+        assert feature_function in featurizer_config[index + window_center]
+
+    feature_vec = seq_vec.toarray()
+    tokens = test_message.get(TOKENS_NAMES[TEXT])
+
+    assert len(tokens) == feature_vec.shape[0]
+
+    # For each word, check that the features we expect are the features we received
+    for i, _ in enumerate(tokens):
+        token_features = feature_vec[i, :]
+
+        # For each expected feature, check that it's 1, then zero it out
+        for j, position_features in enumerate(featurizer_config):
+            position = i + j - window_center
+            if position >= 0 and position < len(tokens):
+                for feature_function in position_features:
+                    feature_name = f"{j - window_center}:{feature_function}"
+                    feature_value_str = str(
+                        featurizer.function_dict[feature_function](tokens[position])
+                    )
+                    feature_idx = featurizer.feature_to_idx_dict[feature_name][
+                        feature_value_str
+                    ]
+                    assert token_features[feature_idx] == 1.0
+                    token_features[feature_idx] = 0.0
+
+        # Since all expected features were zeroed out, result should be all zeros
+        assert not np.any(token_features)
```

Comment:
I didn't fully understand what this test is doing but I think this should ideally be a separate test. Also, there is a lot of "logic" inside the test on what the code should be doing and it's a bit hard to follow. Are we testing a particular edge case here?

Reply:
Just so I understand: do you mean that this should be separated out from the rest of the test?

In a nutshell, the this test is checking that the features returned for the test input are 1 if and only if they were features requested in `featurizer_config`. Rather than use a hardcoded expected features (see comment above), it does this by indexing the features by name via `feature_to_idx_dict`. This is arguably a whitebox test, but afaict `feature_to_idx_dict` is public so it seems ok to test it.

Reply:
As I said above, having this as a complete test is still useful, so let's keep it.